File: cracking_the_coding_interview_6th/chapter16/cpp/16_25_v01.cpp
#include <exception>
#include <fstream>
#include <map>
#include <memory>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>
using namespace std;
using ll = long long;

class EmptyListException : public exception {
  virtual const char *what() const throw() {
    return "list is empty";
  }
} ele;

class FullListException : public exception {
  virtual const char *what() const throw() {
    return "list is full";
  }
} fle;

template <typename T> struct Node {
  string key;
  T value;
  Node() {}
  Node(string _key, T _value) : key(_key), value(_value) {}
  shared_ptr<Node<T>> prv = nullptr;
  shared_ptr<Node<T>> nxt = nullptr;
  string repr() {
    stringstream ss;
    ss << "(" << key << ", " << value << ")";
    return ss.str();
  }
};

template <typename T> class LRUCache {
  ll _capacity;
  ll _size = 0;
  shared_ptr<Node<T>> _head = nullptr;
  shared_ptr<Node<T>> _tail = nullptr;
  map<string, shared_ptr<Node<T>>> _m;
  void prepend(shared_ptr<Node<T>> node) {
    if (!_head) {
      _head = _tail = node;
      return;
    }
    shared_ptr<Node<T>> h = _head;
    _head = node;
    _head->nxt = h;
    h->prv = _head;
  }
  void touch(shared_ptr<Node<T>> node) {
    if (node == _head) return;
    del(node);
    prepend(node);
  }
  void del(shared_ptr<Node<T>> node) {
    if (!_head) throw ele;
    if (node == _head) _head = node->nxt;
    if (node == _tail) _tail = node->prv;
    if (node->nxt) { node->nxt->prv = node->prv; }
    if (node->prv) { node->prv->nxt = node->nxt; }
  }

public:
  LRUCache(ll capacity) {
    if (capacity <= 0)
      throw invalid_argument{"capacity must be greater than 0."};
    _capacity = capacity;
  }
  ll size() {
    return _size;
  }
  void add(const string &key, const T &value) {
    shared_ptr<Node<T>> node = make_shared<Node<T>>(key, value);
    if (_m.find(key) == _m.end() && _size == _capacity) {
      del(_tail);
      prepend(node);
      return;
    }
    prepend(node);
    _m[key] = node;
    ++_size;
  }
  T get(const string &key, const T &default_value) {
    try {
      return operator[](key);
    } catch (const invalid_argument &ia) { return default_value; }
  }
  T operator[](const string &key) {
    if (_m.find(key) == _m.end())
      throw invalid_argument{"key=" + key + " not found"};
    shared_ptr<Node<T>> node = _m[key];
    touch(node);
    return node->value;
  }
  string repr() {
    shared_ptr<Node<T>> node = _head;
    string res;
    while (node) {
      res += node->repr();
      res += ' ';
      node = node->nxt;
    }
    if (res.size()) res.pop_back();
    return res;
  }
};

int main() {
  LRUCache<ll> c{5};
  c.add("a", 1);
  c.add("b", 2);
  c.add("c", 3);
  puts(c.repr().c_str()); // (c, 3) (b, 2) (a, 1)
  c["a"];
  puts(c.repr().c_str()); // (a, 1) (c, 3) (b, 2)
  c.add("e", 5);
  c.add("f", 6);
  puts(c.repr().c_str()); // (f, 6) (e, 5) (a, 1) (c, 3) (b, 2)
  c.add("g", 7);
  puts(c.repr().c_str()); // (g, 7) (f, 6) (e, 5) (a, 1) (c, 3)
  c.add("h", 8);
  puts(c.repr().c_str()); // (h, 8) (g, 7) (f, 6) (e, 5) (a, 1)
  c["a"];
  c.add("i", 9);
  puts(c.repr().c_str()); // (i, 9) (a, 1) (h, 8) (g, 7) (f, 6)
}

File: cracking_the_coding_interview_6th/chapter16/python/16_25_v01.py
from __future__ import annotations
from typing import TypeVar, Generic
from collections import UserDict

T = TypeVar('T')


class EmptyListException(Exception):
    pass


class Node(Generic[T]):
    def __init__(self, key: int, value: T):
        self.key: int = key
        self.value: T = value
        self.prv: Node = None
        self.nxt: Node = None

    def __repr__(self) -> str:
        return f"({self.key}, {self.value})"


class LRUCache(Generic[T], UserDict):
    def __init__(self, capacity: int):
        if capacity <= 0:
            raise ValueError("capacity must be greater than 0.")
        super().__init__()
        self.__capacity: int = capacity
        self.__head: Node = None
        self.__tail: Node = None

    def __repr__(self) -> str:
        node = self.__head
        res = []
        while node is not None:
            res.append(repr(node))
            node = node.nxt
        return ' '.join(res)

    def __getitem__(self, key: int) -> T:
        if key not in self:
            raise KeyError(f"key={key} does not exist.")
        node = self.data[key]
        self.__touch(node)
        return node.value

    def __setitem__(self, key: int, value: T) -> None:
        node = Node(key, value)
        if key not in self and len(self) == self.__capacity:
            self.__del(self.__tail)
            self.__prepend(node)
            return
        self.__prepend(node)
        self.data[key] = node

    def __prepend(self, node: Node[T]) -> None:
        if self.__head is None:
            self.__head = self.__tail = node
            return
        h = self.__head
        self.__head = node
        self.__head.nxt = h
        h.prv = self.__head

    def __del(self, node: Node[T]) -> None:
        if self.__head is None:
            raise EmptyListException("list is empty.")
        if node is self.__head:
            self.__head = node.nxt
        if node is self.__tail:
            self.__tail = node.prv
        if node.nxt is not None:
            node.nxt.prv = node.prv
        if node.prv is not None:
            node.prv.nxt = node.nxt

    def __touch(self, node: Node[T]) -> None:
        if node is self.__head:
            return
        self.__del(node)
        self.__prepend(node)

    def get(self, key: int, default_value: T) -> T:
        try:
            return self[key]
        except KeyError:
            return default_value


c = LRUCache(5)
c[1] = "a"
c[2] = "b"
c[3] = "c"
print(c)    # (3, c) (2, b) (1, a)
c[1]
print(c)    # (1, a) (3, c) (2, b)
c[4] = "d"
c[5] = "e"
print(c)    # (5, e) (4, d) (1, a) (3, c) (2, b)
c[6] = "f"
print(c)    # (6, f) (5, e) (4, d) (1, a) (3, c)
c[7] = "g"
print(c)    # (7, g) (6, f) (5, e) (4, d) (1, a)
c[1]
c[8] = "h"
print(c)    # (8, h) (1, a) (7, g) (6, f) (5, e)
